python/src/iceberg/io/pyarrow.py

```diff
@@ -0,0 +1,154 @@
+
+        Args:
+            overwrite(bool): Whether to overwrite the file if it already exists
+
+        Returns:
+            pyarrow.lib.NativeFile: A NativeFile instance for the file located at self.location
+
+        Raises:
+            FileExistsError: If the file already exists at `self.location` and `overwrite` is False
+        """
+        if not overwrite and self.exists:
+            raise FileExistsError(
+                f"A file already exists at this location. If you would like to overwrite it, set `overwrite=True`: {self.location}"
+            )
+        output_file = self._filesystem.open_output_stream(self._path)
+        if not isinstance(output_file, OutputStream):
+            raise TypeError(
+                f"Object of type {type(output_file)} returned from PyArrowFile.create(...) does not match the OutputStream protocol."
+            )
+        return output_file
+
+    def to_input_file(self) -> "PyArrowFile":
+        """Returns a new PyArrowFile for the location of an existing PyArrowFile instance
+
+        This method is included to abide by the OutputFile abstract base class. Since this implementation uses a single
+        PyArrowFile class (as opposed to separate InputFile and OutputFile implementations), this method effectively returns
+        a copy of the same instance.
+        """
+        return PyArrowFile(self.location)
+
+
+class PyArrowFileIO(FileIO):
+    def new_input(self, location: str) -> PyArrowFile:
+        """Get a PyArrowFile instance to read bytes from the file at the given location
+
+        Args:
+            location(str): A URI or a path to a local file
+
+        Returns:
+            PyArrowFile: A PyArrowFile instance for the given location
+        """
+        return PyArrowFile(location)
+
+    def new_output(self, location: str) -> PyArrowFile:
+        """Get a PyArrowFile instance to write bytes to the file at the given location
+
+        Args:
+            location(str): A URI or a path to a local file
+
+        Returns:
+            PyArrowFile: A PyArrowFile instance for the given location
+        """
+        return PyArrowFile(location)
+
+    def delete(self, location: Union[str, InputFile, OutputFile]) -> None:
+        """Delete the file at the given location
+
+        Args:
+            location(str, InputFile, OutputFile): The URI to the file--if an InputFile instance or an
+            OutputFile instance is provided, the location attribute for that instance is used as the location
+            to delete
+        """
+        str_path = location.location if isinstance(location, (InputFile, OutputFile)) else location
+        filesystem, path = FileSystem.from_uri(str_path)  # Infer the proper filesystem
+        try:
+            filesystem.delete_file(path)
+        except OSError:
+            raise FileNotFoundError(f"File could not be deleted because it does not exist: {str_path}")
```

Comment:
Is it possible to pass the original exception as a cause? This assumes that any `OSError` means that the file doesn't exist, but that doesn't seem correct to me.

Reply:
I think this goes back to the inconsistency I described in my [other reply](https://github.com/apache/iceberg/pull/4081#discussion_r810664650). We could raise a more generic exception here and the user can see the more detailed pyarrow exception in the traceback:
```py
from pyarrow.fs import FileSystem

def delete(location):
    try:
        fs, path = FileSystem.from_uri(location)
        fs.delete_file(path)
    except Exception:
        raise RuntimeError(f"Cannot delete file: {location}")

delete("s3://foo/bar.txt")
```
output:
```py
OSError                                   Traceback (most recent call last)
<ipython-input-7-fd79114b19ff> in foo()
      3         fs, path = FileSystem.from_uri("s3://foo/bar.txt")
----> 4         fs.delete_file(path)
      5     except Exception:

/usr/local/lib/python3.9/site-packages/pyarrow/_fs.pyx in pyarrow._fs.FileSystem.delete_file()

/usr/local/lib/python3.9/site-packages/pyarrow/error.pxi in pyarrow.lib.check_status()

OSError: When getting information for key 'bar.txt' in bucket 'foo': AWS Error [code 15]: No response body.

During handling of the above exception, another exception occurred:

RuntimeError                              Traceback (most recent call last)
<ipython-input-8-c19b6d9633cf> in <module>
----> 1 foo()

<ipython-input-7-fd79114b19ff> in foo()
      4         fs.delete_file(path)
      5     except Exception:
----> 6         raise RuntimeError("Cannot delete file: s3://foo/bar.txt")
      7 
      8 

RuntimeError: Cannot delete file: s3://foo/bar.txt
```

Reply:
Rather than RuntimeError, I'd use a more specific error type, but I like the try/except block to normalize this.

For OSError specifically, I think we also need to see what kind of OSError it is and allow it through if it isn't the one that means the file is missing.

Reply:
Yeah, an [OSError](https://docs.python.org/3/library/exceptions.html#OSError) just passes the [C errno](https://man7.org/linux/man-pages/man3/errno.3.html) through. Here, you'll need to check for `ENOENT`, which is no such file or directory.

That probably also means that for the `open_output_stream` you can catch `OSError` and check `errno` for `EEXIST`, which indicates that the file exists.

Reply:
Digging some more I've found some more diverging behavior. For local files, `e.errno` is 2 as expected to represent `ENOENT` when the file does not exist.

For S3 though, there's no value set at all for `e.errno`. Instead:

- If the issue is access denied to the bucket, it hits this [ErrorToStatus](https://github.com/apache/arrow/blob/master/cpp/src/arrow/filesystem/s3_internal.h#L112) function that returns an `OSError` with nothing but a string description containing "AWS Error [code 15]" which corresponds to [ACCESS_DENIED](https://sdk.amazonaws.com/cpp/api/LATEST/_s3_errors_8h_source.html) in the aws cpp sdk.
- If the bucket access is fine but the resource doesn't exist, it hits this [PathNotFound](https://github.com/apache/arrow/blob/master/cpp/src/arrow/filesystem/s3fs.cc#L2143) error which returns an `OSError` with nothing but a string description containing "Path does not exist: foo".

I'm not enthusiastic about error handling based on substring checks, but the only thing I can think of here is checking for either `e.errno=2` or an error message containing "Path does not exist" in order to appropriately raise a `FileNotFoundError`. That feels fragile but should handle local and s3 files. I haven't checked the behavior for HDFS files but that could also be different.

Reply:
Updated the PR with this approach. The `exists`, `open`, and `create` methods now catch and appropriately raise errors such as `PermissionError`, `FileExistsError`, or `FileNotFoundError`.

For local files this is straightforward and checks `e.errno`, but for S3 errors, it checks the string response. If the error raised can't be determined, the underlying pyarrow error is just raised directly. Even when it's caught, the underlying pyarrow error is always available in the traceback.

I've also added a handful of tests to test for each of these scenarios and added docs for all of them in the `Raises:` section of the method docstrings.